### PR TITLE
Compile GAP functions to Julia - WIP

### PR DESCRIPTION
**Pull request by sebasguts**
This is a WIP proof-of-concept implementation for translating GAP functions into Julia functions, to be able to call them on Julia data types. See example below:

Currently, only some GAP expressions/statements are converted, so this is WIP.

```
gap> bubble_sort := function( list, len )
>     local i, j, tmp;
>     
>     for i in [ 1 .. len ] do
>         for j in [ 1 .. len - 1 ] do
>             if list[ j ] > list[ j + 1 ] then
>                 tmp := list[ j + 1 ];
>                 list[ j + 1 ] := list[ j ];
>                 list[ j ] := tmp;
>             fi;
>         od;
>     od;
>     return list;
> end;
function( list, len ) ... end
gap> 
gap> 
gap> bubble_sort_jl := TRANSPILE_FUNC( bubble_sort );
<Julia: :(function (localvar_1, localvar_2)
      for localvar_3 = 1:localvar_2
          for localvar_4 = 1:localvar_2 - 1
              if localvar_1[localvar_4] > localvar_1[localvar_4 + 1]
                  localvar_5 = localvar_1[localvar_4 + 1]
                  localvar_1[localvar_4 + 1] = localvar_1[localvar_4]
                  localvar_1[localvar_4] = localvar_5
              end
          end
      end
      return localvar_1
  end)>
gap> bubble_sort_jl := Julia.Base.eval( bubble_sort_jl );
<Julia: getfield(Base, Symbol("##738#739"))()>
gap> 
gap> length := 5000;
5000
gap> 
gap> xx := List( [ 1 .. length ], i -> RandomList( [ 1 .. length ] ) );;
gap> xx_jl := GAPToJulia( JuliaEvalString( "Vector{Int64}"), xx );;
gap> 
gap> bubble_sort( xx, length );;
gap> time;
784
gap> 
gap> bubble_sort_jl( xx_jl, length );;
gap> bubble_sort_jl( xx_jl, length );;
gap> time;
16
```

### CODE REVIEW

File: pkg/GAPJulia/JuliaInterface/gap/JuliaTranspiler.gi [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+BindGlobal( "TRANSPILE_FUNC",
+  function( func )
+    local nr_args, arg_expr, exprs, tree;
+    tree := SYNTAX_TREE( func );
+    nr_args := tree!.narg;
+    arg_expr := List( [ 1 .. nr_args ], SYMBOL_LVAR );
+    arg_expr := CallFuncList( CREATE_EXPRESSION1, Concatenation( [ "tuple" ], arg_expr ) );
+    exprs := [ "function", arg_expr, DISPATCH_EXPRESSION( tree.stats ) ];
+    return CallFuncList( CREATE_EXPRESSION1, exprs );
+end );
```
Comment (fingolfin) [inline pkg/GAPJulia/JuliaInterface/gap/JuliaTranspiler.gi]: Add missing newline at the end?

File: pkg/GAPJulia/JuliaInterface/gap/JuliaTranspiler.gi [hunk 2]
```diff
@@ -0,0 +1,262 @@
+#############################################################################
+##
+##  JuliaInterface package
+##
+##  Copyright 2019
+##    Thomas Breuer, RWTH Aachen University
+##    Sebastian Gutsche, Siegen University
```
Comment (fingolfin) [inline pkg/GAPJulia/JuliaInterface/gap/JuliaTranspiler.gi]: That's a bit odd. Either list just your name, or else list all three names? Or we switch to an approach like what we now use in the GAP code, and don't list names here at all.

File: pkg/GAPJulia/JuliaInterface/example/transpile.gi [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+xx := List( [ 1 .. length ], i -> RandomList( [ 1 .. length ] ) );;
+xx_jl := GAPToJulia( JuliaEvalString( "Vector{Int64}"), xx );;
+
+bubble_sort( xx, length );;
+time;
+
+bubble_sort_jl( xx_jl, length );;
+bubble_sort_jl( xx_jl, length );;
+time;
```
Comment (fingolfin) [inline pkg/GAPJulia/JuliaInterface/example/transpile.gi]: Add trailing newline

**Comment by sebasguts** [comment]
This should be replaced completely. I have some ideas on how to do it, and one of the ideas involve completely write it in Julia. Lets just keep this open as a reminder, but not worry about it for now.

**Pull request closed without merging**